sdks/java/core/src/main/java/org/apache/beam/sdk/coders/ZstdCoder.java

```diff
@@ -0,0 +1,110 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.coders;
+
+import com.github.luben.zstd.Zstd;
+import com.github.luben.zstd.ZstdCompressCtx;
+import com.github.luben.zstd.ZstdDecompressCtx;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.List;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.util.CoderUtils;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.ImmutableList;
+
+/**
+ * Wraps an existing coder with Zstandard compression. It makes sense to use this coder when it's
+ * likely that the encoded value is quite large and compressible or when a dictionary is available
+ * to improve compression performance.
+ */
+public class ZstdCoder<T> extends StructuredCoder<T> {
+  private final Coder<T> innerCoder;
+  private final @Nullable byte[] dict;
+  private final int level;
+
+  /** Wraps the given coder into a {@link ZstdCoder}. */
+  public static <T> ZstdCoder<T> of(Coder<T> innerCoder, byte[] dict, int level) {
+    return new ZstdCoder<>(innerCoder, dict, level);
+  }
+
+  /** Wraps the given coder into a {@link ZstdCoder}. */
+  public static <T> ZstdCoder<T> of(Coder<T> innerCoder, byte[] dict) {
+    return new ZstdCoder<>(innerCoder, dict, Zstd.defaultCompressionLevel());
+  }
+
+  /** Wraps the given coder into a {@link ZstdCoder}. */
+  public static <T> ZstdCoder<T> of(Coder<T> innerCoder, int level) {
+    return new ZstdCoder<>(innerCoder, null, level);
+  }
+
+  /** Wraps the given coder into a {@link ZstdCoder}. */
+  public static <T> ZstdCoder<T> of(Coder<T> innerCoder) {
+    return new ZstdCoder<>(innerCoder, null, Zstd.defaultCompressionLevel());
+  }
+
+  private ZstdCoder(Coder<T> innerCoder, @Nullable byte[] dict, int level) {
+    this.innerCoder = innerCoder;
+    this.dict = dict;
+    this.level = level;
+  }
+
+  @Override
+  public void encode(T value, OutputStream os) throws IOException {
+    ZstdCompressCtx ctx = new ZstdCompressCtx();
+    try {
+      ctx.setLevel(level);
+      ctx.setMagicless(true); // No magic since we know this will be compressed data on decode.
+      ctx.setDictID(false); // No dict ID since we initialize the coder with the expected dict.
+      ctx.loadDict(dict);
+
+      byte[] encoded = CoderUtils.encodeToByteArray(innerCoder, value);
+      ctx.setPledgedSrcSize(encoded.length);
+      byte[] compressed = ctx.compress(encoded);
+      ByteArrayCoder.of().encode(compressed, os);
+    } finally {
+      ctx.close();
+    }
+  }
+
```

Comment:
Will do, this follows from wrongly subclassing StructuredCoder.

Reply:
@lukecwik I think structuralValue could also delegate to innerCoder.structuralValue to skip compression when consistentWithEquals returns false, what do you think?

Reply:
SnappyCoder should do the same as you pointed out.

The issue with structuralValue is if that consistentWithEquals is false you don't know what the inner coder will do (will it encode it or wrap the value to make equals work out). If it is the former then compressing it means that we pass around smaller byte arrays in memory and if it is the latter then we would want to delegate it since we don't want to encode it. I think your right though that the latter is better and we should delegate it since we would want that behavior when we can get it and passing around the uncompressed byte arrays albeit annoying isn't as big of a loss.

Reply:
Resolved in af74511.